pallets/parachain-staking/src/auto_compound.rs

```diff
@@ -0,0 +1,383 @@
+use sp_runtime::Percent;
+use sp_std::prelude::*;
+use sp_std::vec::Vec;
+
+/// Represents the auto-compounding amount for a delegation.
+#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo, PartialOrd, Ord)]
+pub struct AutoCompoundConfig<AccountId> {
+	pub delegator: AccountId,
+	pub value: Percent,
+}
+
+/// Represents the auto-compounding [Delegations] for `T: Config`
+#[derive(Clone, Eq, PartialEq, RuntimeDebug)]
+pub struct AutoCompoundDelegations<T: frame_system::Config> {
+	inner: Vec<AutoCompoundConfig<T::AccountId>>,
+	_phantom: PhantomData<T>,
+}
+
+impl<T> AutoCompoundDelegations<T>
+where
+	T: Config,
+{
+	/// Creates a new instance of [AutoCompoundingDelegations].
+	pub fn new(delegations: Vec<AutoCompoundConfig<T::AccountId>>) -> Self {
+		Self {
+			inner: delegations,
+			_phantom: Default::default(),
+		}
+	}
+
+	/// Retrieves an instance of [AutoCompoundingDelegations] storage as [AutoCompoundDelegations].
+	pub fn get_storage(candidate: &T::AccountId) -> Self {
+		Self {
+			inner: <AutoCompoundingDelegationsStorage<T>>::get(candidate),
+			_phantom: Default::default(),
+		}
+	}
+
+	/// Inserts the current state to [AutoCompoundingDelegations] storage.
+	pub fn set_storage(self, candidate: &T::AccountId) {
+		<AutoCompoundingDelegationsStorage<T>>::insert(candidate, self.inner)
+	}
+
+	/// Retrieves the auto-compounding value for a delegation. The `delegations_config` must be a
+	/// sorted vector for binary_search to work.
+	pub fn get_for_delegator(&self, delegator: &T::AccountId) -> Option<Percent> {
+		match self.inner.binary_search_by(|d| d.delegator.cmp(&delegator)) {
+			Ok(index) => Some(self.inner[index].value.clone()),
+			Err(_) => None,
+		}
+	}
+
+	/// Sets the auto-compounding value for a delegation. The `delegations_config` must be a sorted
+	/// vector for binary_search to work.
+	pub fn set_for_delegator(&mut self, delegator: T::AccountId, value: Percent) -> bool {
+		match self.inner.binary_search_by(|d| d.delegator.cmp(&delegator)) {
+			Ok(index) => {
+				if self.inner[index].value == value {
+					false
+				} else {
+					self.inner[index].value = value;
+					true
+				}
+			}
+			Err(index) => {
+				self.inner
+					.insert(index, AutoCompoundConfig { delegator, value });
+				true
+			}
+		}
+	}
+
+	/// Removes the auto-compounding value for a delegation.
+	/// Returns `true` if the entry was removed, `false` otherwise. The `delegations_config` must be a
+	/// sorted vector for binary_search to work.
+	pub fn remove_for_delegator(&mut self, delegator: &T::AccountId) -> bool {
+		match self.inner.binary_search_by(|d| d.delegator.cmp(&delegator)) {
+			Ok(index) => {
+				self.inner.remove(index);
+				true
+			}
+			Err(_) => false,
+		}
+	}
+
+	/// Returns the length of the inner vector.
+	pub fn len(&self) -> u32 {
+		self.inner.len() as u32
+	}
+
+	/// Returns a reference to the inner vector.
+	pub fn inner(&self) -> &Vec<AutoCompoundConfig<T::AccountId>> {
```

Comment:
This function seem to be used only for tests, can you add `#[cfg(test)]` attributes?